api/src/wfl/sink.clj

```diff
@@ -317,27 +325,63 @@
                                  {:job      job
                                   :workflow workflow})))))))
 
+;; visible for testing
+(defn rename-gather-bulk
+  "Transform the `values` using the transformation defined in `mapping`, building bulk
+   load file models instead of strings."
+  ([workflow-id dataset table values mapping]
+   (rename-gather-bulk workflow-id dataset table values mapping ""))
+  ([workflow-id {:keys [schema] :as dataset} table values mapping target-bucket]
+   (letfn [(literal? [x] (str/starts-with? x "$"))
+           (datatype [k] (let [columns (->> schema
+                                            :tables
+                                            (filter #(= table (:name %)))
+                                            first
+                                            :columns)]
+                           (-> (filter #(= k (:name %)) columns)
+                               first
+                               :datatype)))
+           (fileref? [k] (= (datatype (name k)) "fileref"))
+           (boolean? [k] (= (datatype (name k)) "boolean"))
+           (get-target [url]
+             (let [[bucket obj] (storage/parse-gs-url url)]
+               (if (= target-bucket "")
+                 (storage/gs-url bucket (str/join "/" [workflow-id (util/basename obj)]))
+                 (str target-bucket workflow-id "/" (util/basename obj)))))
+           (go! [k v]
+             (cond (fileref? k) (let [val (values (keyword v))]
+                                  [k {:description (util/basename val)
+                                      :mimeType   (mime-type/ext-mime-type val)
+                                      :sourcePath val
+                                      :targetPath (get-target val)}])
+                   (literal? v) [k (subs v 1 (count v))]
+                   (boolean? v) [k (values (keyword v))]
+                   (string?  v) [k (values (keyword v))]
+                   (map?     v) [k (rename-gather-bulk workflow-id dataset table values v)]
+                   (coll?    v) [k (keep (go! k v))]
+                   :else        (throw (ex-info "Unknown operation"
+                                                {:operation v}))))]
+     (into {} (for [[k v] mapping] (go! k v))))))
```

Comment:
Why doesn't `rename-gather` have `boolean` and `fileref` clauses?
Do we still need `rename-gather`?  Looks like it.
(-I'm so confused!-)

Reply:
The `rename-gather` function is used for Terra Workspace Sinks, which it only sends the fileref fields as strings. In this case, for TerraDataRepoSink's single step ingest, we need to send a json blob with the sourcePath, targetPath, mimeType and a description. Because of this, there is special handling to turn the string based fileref outputs to this BulkFileLoadModel that TDR expects.

Reply:
But shouldn't `rename-gather` be a relational subset of `rename-gather-bulk`?
If not, maybe we should rename them for `workspace` vs `tdr` sinks?

Reply:
Must `target-bucket` name a bucket?
It looks like a generic GCS prefix or "folder".

Reply:
And now I'm having trouble seeing how that can be used in practice.
What sets target-bucket? I see it used in a test.
Is it intended for production?

Reply:
It was admittedly added so that I could save the files to a specific bucket in the test so that the files could be cleaned up by the fixture afterwards. But could serve a purpose in production in the future if there is a preference on where the target files are saved during ingest.